Frontend/nix/watch.nix

```diff
@@ -0,0 +1,115 @@
+{ ... }:
+{
+  imports = [
+    ../node
+  ];
+
+  perSystem = { config, self', lib, pkgs, system, ... }:
+    let
+      webpack-dev-server = pkgs.writeShellApplication {
+        name = "webpack-dev-server-wrapped";
+        runtimeInputs = [ self'.packages.nodejs ];
+        text = ''
+          export NODE_PATH="${self'.packages.nodeDependencies}/node_modules"
+          PIPE="$1"
+          shift
+          tail -n1 -f "$PIPE" | node ${./watch.js} "$@"
+        '';
+      };
+
+      purifix-watch =
+        pkgs.writeShellApplication {
+          name = "purifix-watch";
+          runtimeInputs = [
+            pkgs.watchexec
+          ] ++ config.packages.ui-common.develop.buildInputs;
+          text = ''
+            PIPE=$1
+            shift
+            SRC_FOLDERS=()
+            SRC_GLOBS=()
+            for SRC_FOLDER in "$@"; do
+              if [ -d "$SRC_FOLDER/src" ]; then
+                SRC_FOLDERS+=( -w "$SRC_FOLDER/src" )
+                SRC_GLOBS+=( "\"$SRC_FOLDER/src/**/*.purs\"" )
+              fi
+            done
+            watchexec "''${SRC_FOLDERS[@]}" -r -f "**/*.purs" -- \
+              "echo 'suspend' >$PIPE;purifix ''${SRC_GLOBS[*]} && echo 'ready' >$PIPE"
+          '';
+        };
```

Comment:
buildInputs is to support running outside of the devShell, I have removed it as requested.

I disagree that this is really so decoupled from the project its self. The fact is most people are not going to want to use a pipe to communicate between two different watch processes. This whole two processes intertwined thing is really more of a hack necessitated by the way the existing project was using webpack. In an ideal world we could just use the existing webpack loader but I was unable to get it to work correctly with purifix. 

In general the reason this whole thing feels like its all intertwined is because it is inherently and breaking apart things that are inherently intertwined will not make anything clearer. The webpack-dev-server process and the purifix-watch process are engaged in a very specific dance where they are trading off work, they are necessarily decoupled via a pipe which has a fairly simple protocol but each piece independently makes no sense.

I also don't really believe that purifix will ever directly solve this problem. There are two things going on here that are usually not relevant. 
1. We have a multi-project repo where two of the packages are mutually exclusive due to overlapping namespacing. This is why we even need to do the globbing/src watching instead of just using purifix-project which will build your whole project instead of just acting like the `purs compile` command.
2. We are using webpack in a fairly abnormal way on the project. Webpack watch doesn't seem to debounce the changes coming from purescript well enough on its own so it will constantly recompile unless told to suspend its self. 

The fact is this whole long lived independent watch processes stuff is a hack necessitated by the slowness of webpack combined with the inability to integrate directly in a sensible way.

Reply:
The whole `purfix-watch` + (your custom) `webpack-dev-server` combination is [an area of volatility](https://www.informit.com/articles/article.aspx?p=2995357&seqNum=2) that NY frontend devs shouldn't have to concern themselves with when reading top-level Nix. When better solutions arise in future, we should be able to swap it in easily. The interface here really just says "take this directory, and compile it with common directory, watch for it as part of webpack dev server".

> In an ideal world we could just use the existing webpack loader but I was unable to get it to work correctly with purifix.

Yea, but why does it have to be an *existing* loader? I am doing some reading on how on webpack works, and it seems to me that we are not approaching this problem cleanly: we're writing custom hack when we could just use webpack's loader mechanism. Have you actually considered writing a _new_ (or reusing an existing) loader that works with our purifix project? 

> In an ideal world we could just use the existing webpack loader but I was unable to get it to work correctly with purifix.

By "the existing webpack loader" are you referring to https://github.com/ethul/purs-loader? What's the problem exactly? Can we extend it to work with purifix? I imagine this work would be useful beyond nammayatri, and in other PureScript projects anyway.

(As for overlapping namespacing, I thin we should fix it separately if it turns out to be a problem).

Reply:
>The whole purfix-watch + (your custom) webpack-dev-server combination is [an area of volatility](https://www.informit.com/articles/article.aspx?p=2995357&seqNum=2) that NY frontend devs shouldn't have to concern themselves with when reading top-level Nix.
How is nix/webpack-dev-server/default.nix and nix/purifix-watch.nix any more or less top level than nix/watch.nix?

> Have you actually considered writing a new (or reusing an existing) loader that works with our purifix project?
Yes I already attempted this and determined that the amount of effort to make a new loader would be fairly significant and so disregarded it as a practical option.

> By "the existing webpack loader" are you referring to https://github.com/ethul/purs-loader? What's the problem exactly?
It fails to properly rewrite the imports post compilation. The main issue with webpack and purescript working together is that they have different expectations about their roles. Webpack assumes that it has a compiler that can convert a single file at a time to js. Purescript wants to compile everything all at once to js. To properly handle webpacks view where it wants to collect all the dependencies itsself the purs-loader rewrites the output js files to point back to the purs files. However, in our setup it appears to fail to properly determine where the src is located. In this process it mangles the file and causes the whole process to crash. 

I can try again to get it working but it really does seem like were getting into pareto principle territory over here.

Reply:
`nix/watch.nix` contains two things: the process-compose/mission-compose configuration (which qualifies as top-level Nix that devs are expected to modify) as well as the implementation details of this watch hack (which doesn't need to appear there).

> To properly handle webpacks view where it wants to collect all the dependencies itsself the purs-loader rewrites the output js files to point back to the purs files. However, in our setup it appears to fail to properly determine where the src is located. In this process it mangles the file and causes the whole process to crash.

I'm not sure if I completely understood that. In generally, whenever something like this happens the course of action I'd recommend is:

1. Open an upstream issue detailing the issue (in this case, on https://github.com/ethul/purs-loader)
2. Attempt to fix the issue and open a upstream PR
3. If the fix is difficult or impossible, explain so in that upstream issue, and then link all that information in the local hack (webpack-dev-server/ purifix-watch) so that future engineers modifying this Nix will have a context as to why all of this exists.

> I can try again to get it working but it really does seem like were getting into pareto principle territory over here.

Yea, please do. It is ideal to fix https://github.com/ethul/purs-loader itself - but [timebox](https://en.wikipedia.org/wiki/Timeboxing) the whole thing, as we don't want to be spending too much time there. Do open the upstream issue and link to it if that effort fails.